### PR TITLE
Add ability to use built-in `pickle` for saving AutoMLSearch

### LINKED ISSUE #2174: Ability to pickle AutoMLSearch
Our `AutoMLSearch.save` and `load` methods support serializing/deserializing `AutoMLSearch` instances using `cloudpickle`. This issue supports doing the same using python's `pickle`.

The pickling method could be an argument to `save` and `load`. Note that whatever pattern we use here, we should use the same pattern elsewhere in our code, like for `PipelineBase.save` and `load`.

### PR DESCRIPTION
**Pull request by christopherbunn**
Resolves #2174

### CODE REVIEW

File: evalml/automl/automl_search.py [hunk 1]
```diff
@@ -1331,7 +1347,7 @@ def load(file_path):
             AutoSearchBase object
         """
         with open(file_path, "rb") as f:
-            return cloudpickle.load(f)
+            return pickle.load(f)
```
Comment (freddyaboulton) [inline evalml/automl/automl_search.py:1350]: So we can save with cloudpickle and read with pickle?! Wow, I had no idea that works hehe.

I feel like we should accept an argument here for the "pickle type"? Feels weird to offer a choice of library for `save` but not respect that in `load`. Not blocking though.

**Review by freddyaboulton — APPROVED** [review]
@christopherbunn Looks good! The original issue also mentions doing something similar for pipelines. Is the plan is to do that in this PR or a follow up?

File: evalml/automl/automl_search.py [hunk 2]
```diff
@@ -1307,18 +1308,33 @@ def best_pipeline(self):
...
+        elif pickle_type == "pickle":
+            pkl_lib = pickle
+        else:
+            raise ValueError(
+                f"`pickle_type` must be either 'pickle' or 'cloudpickle'. Received {pickle_type}"
+            )
+
         with open(file_path, "wb") as f:
-            cloudpickle.dump(self, f, protocol=pickle_protocol)
+            pkl_lib.dump(self, f, protocol=pickle_protocol)
```
Comment (angela97lin) [inline evalml/automl/automl_search.py:1337]: What if the user uses a cloudpickle protocol while trying to use pickle? 🤔

**Review by angela97lin — COMMENTED** [review]
I'm trying to pickle using a jupyter notebook and I'm getting the following, am I doing something wrong?:

```
automl_ = AutoMLSearch(X, y, problem_type="regression")
automl_.search()
automl_.save("test.pkl", pickle_type="pickle")
```


I get:
```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-12-e1254c6fb959> in <module>
----> 1 automl_.save("test.pkl", pickle_type="pickle")

~/Desktop/evalml/evalml/automl/automl_search.py in save(self, file_path, pickle_type, pickle_protocol)
   1335 
   1336         with open(file_path, "wb") as f:
-> 1337             pkl_lib.dump(self, f, protocol=pickle_protocol)
   1338 
   1339     @staticmethod

TypeError: can't pickle module objects
```

It works if I use cloudpickle 🤔

File: evalml/automl/automl_search.py (hunk 2 above)
Comment (christopherbunn) [reply evalml/automl/automl_search.py:1337]: Apparently cloudpickles can be opened by the regular pickling library (according to their example on their [README.md](https://github.com/cloudpipe/cloudpickle) and the doc string for `cloudpickle.py` in the attached screenshot)! I didn't know this before so that's pretty neat.

image

**Comment by freddyaboulton** [comment]
@angela97lin Are you running in a jupyter notebook? 

The following works for me when running from a script:

```python
from evalml.demos import load_diabetes
from evalml.automl import AutoMLSearch

X, y = load_diabetes()

automl = AutoMLSearch(X, y, "regression")
automl.search()
automl.save("automl.pkl", pickle_type="pickle")
automl2 = automl.load("automl.pkl")
```

**Comment by christopherbunn** [comment]
> I'm trying to pickle using a jupyter notebook and I'm getting the following, am I doing something wrong?:
> 
> ```
> automl_ = AutoMLSearch(X, y, problem_type="regression")
> automl_.search()
> automl_.save("test.pkl", pickle_type="pickle")
> ```
> 
> I get:
> 
> ```
> ---------------------------------------------------------------------------
> TypeError                                 Traceback (most recent call last)
> <ipython-input-12-e1254c6fb959> in <module>
> ----> 1 automl_.save("test.pkl", pickle_type="pickle")
> 
> ~/Desktop/evalml/evalml/automl/automl_search.py in save(self, file_path, pickle_type, pickle_protocol)
>    1335 
>    1336         with open(file_path, "wb") as f:
> -> 1337             pkl_lib.dump(self, f, protocol=pickle_protocol)
>    1338 
>    1339     @staticmethod
> 
> TypeError: can't pickle module objects
> ```
> 
> It works if I use cloudpickle 🤔

Hmm, that's so odd! I am able to generate a pickle with the test and through a script version of my test Jupyter notebook. At the same time, I am also running into the same error as you when I try to run it interactively in a Jupyter notebook. Let me look into this...

File: evalml/automl/automl_search.py (hunk 2 above)
Comment (angela97lin) [reply evalml/automl/automl_search.py:1337]: So cool 🤩

**Comment by angela97lin** [comment]
@freddyaboulton @christopherbunn Yup, running via jupyter notebook. What I was curious about was if it was possible to save in a file and load in completely different file so I wanted to have two separate notebooks to test but ran into this, heh.

**Comment by christopherbunn** [comment]
I think I figured out the issue: in the notebook environment, we have the search iteration plot appear. However, in the test and script environment, we aren't showing a search iteration plot. As a result, the jupyter notebook version has the plot stored in `automl.search_iteration_plot` whereas the script version has None in this slot. I am able to pickle the search in Jupyter when i set `automl.search_iteration_plot=None`. 

I'm guessing that for some reason the plot doesn't support pickling? This is a bit odd considering that plotly has support for pickling per [this PR on their repo](https://github.com/plotly/plotly.py/pull/1191). I'll see if I can find a workaround or a deeper root cause in Plotly.

**Comment by angela97lin** [comment]
@christopherbunn Wow, good find! If you can't find a good workaround, I think it could also be fine to say that the plots will not be pickled and set `automl.search_iteration_plot=None` before saving, not sure if others have different opinions on that 🤷‍♀️

**Comment by freddyaboulton** [comment]
I think the issue is that the `PipelineSearchPlot` has a reference to `plotly.graph_objects`, which is a module. Maybe we can refactor that? But I think not saving the plot is also a valid approach.

image

We might want to update the AutoMLSearch User Guide to talk about also using pickle to save automl and the limitations. Would also give us "coverage" for pickling in a jupyter notebook env which our unit tests don't cover.

**Comment by christopherbunn** [comment]
Turns out I was wrong, the Jupyter notebook version uses a `SearchIterationPlot` object whereas the script/test version stores a plotly figure (which is pickleable). The error is raised because the `_go` class variable in `SearchIterationPlot` stores the module representing the `plotly.graph_objects` class.  

I think the best path is to actually clear out the `_go` class variable after the init since this is the only place where this is used. Initial testing show that pickling works once this is done. I'll test a few edge cases but if it looks good I'll push it up.

File: evalml/automl/automl_search.py (hunk 1 above)
Comment (christopherbunn) [reply evalml/automl/automl_search.py:1350]: Hmm yeah I agree, it does feel symmetrical to do so. I think it would be a no-op though since it looks like the doc for cloudpickle just recommends using the standard python pickler for loading.
